app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -2,71 +2,82 @@ package com.lightricks.feedexercise.ui.feed
 
 import androidx.lifecycle.*
 import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.data.FeedRepository
 import com.lightricks.feedexercise.util.Event
+import io.reactivex.BackpressureStrategy
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.schedulers.Schedulers
 import java.lang.IllegalArgumentException
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
-    private val stateInternal: MutableLiveData<State> = MutableLiveData<State>(DEFAULT_STATE)
+open class FeedViewModel(private val repository: FeedRepository) : ViewModel() {
+    private val isLoading: MutableLiveData<Boolean> = MutableLiveData<Boolean>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private var listOffset = 0
+
+    private val feedItems: LiveData<List<FeedItem>> =
+        LiveDataReactiveStreams.fromPublisher(
+            repository.fetchData().toFlowable(BackpressureStrategy.DROP))
 
     fun getIsLoading(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
+        return isLoading
     }
 
     fun getIsEmpty(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
+        return Transformations.map(feedItems) { it.isEmpty() }
     }
 
     fun getFeedItems(): LiveData<List<FeedItem>> {
-        //todo: fix the implementation
-        return MutableLiveData()
+        return Transformations.map(feedItems) {
+            val size = it.size
+            if (size != 0) {
+                val offset = listOffset % size
+                it.takeLast(size - offset) + it.take(offset)
+            } else listOf()
+        }
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        isLoading.postValue(false)
+        networkErrorEvent.postValue(Event(error.toString()))
     }
 
+
+
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
+        isLoading.postValue(true)
+        repository.refresh()
+            .subscribeOn(Schedulers.io())
```

Comment:
Requiring the I/O thread here reveals an assumption about how the repository works. Let's allow the Repository choose the thread itself.